horovod/spark/util/safe_shell_exec.py

```diff
@@ -0,0 +1,148 @@
+    p = psutil.Process(pid)
+
+    # Terminate children gracefully.
+    for child in p.children():
+        try:
+            child.terminate()
+        except psutil.NoSuchProcess:
+            pass
+
+    # Wait for graceful termination.
+    time.sleep(GRACEFUL_TERMINATION_TIME_S)
+
+    # Send STOP to executor shell to stop progress.
+    p.send_signal(signal.SIGSTOP)
+
+    # Kill children recursively.
+    for child in p.children(recursive=True):
+        try:
+            child.kill()
+        except psutil.NoSuchProcess:
+            pass
+
+    # Kill shell itself.
+    p.kill()
+
+
+def forward_stream(src_fd, dst_stream):
+    if hasattr(dst_stream, 'buffer'):
+        # If dst stream is a text buffer, we need to get its binary buffer.
+        dst_stream = dst_stream.buffer
+
+    with os.fdopen(src_fd, 'rb') as src:
+        while True:
+            line = src.readline()
+            if not line:
+                break
+            dst_stream.write(line)
+            dst_stream.flush()
+
+
+def execute(command, env=None, stdout=None, stderr=None):
+    # Make a pipe for the subprocess stdout/stderr.
+    (stdout_r, stdout_w) = os.pipe()
+    (stderr_r, stderr_w) = os.pipe()
+
+    # Make a pipe for notifying the child that parent has died.
+    (r, w) = os.pipe()
+
+    middleman_pid = os.fork()
+    if middleman_pid == 0:
+        # Close unused file descriptors to enforce PIPE behavior.
+        os.close(w)
+        os.setpgid(0, 0)
+
+        executor_shell = subprocess.Popen(command, shell=True, env=env,
+                                          stdout=stdout_w, stderr=stderr_w)
+
+        sigterm_received = threading.Event()
+
+        def set_sigterm_received(signum, frame):
+            sigterm_received.set()
+
+        signal.signal(signal.SIGINT, set_sigterm_received)
+        signal.signal(signal.SIGTERM, set_sigterm_received)
+
+        def kill_executor_children_if_parent_dies():
+            # This read blocks until the pipe is closed on the other side
+            # due to the process termination.
+            os.read(r, 1)
+            terminate_executor_shell_and_children(executor_shell.pid)
+
+        bg = threading.Thread(target=kill_executor_children_if_parent_dies)
+        bg.daemon = True
+        bg.start()
+
+        def kill_executor_children_if_sigterm_received():
+            sigterm_received.wait()
+            terminate_executor_shell_and_children(executor_shell.pid)
+
+        bg = threading.Thread(target=kill_executor_children_if_sigterm_received)
+        bg.daemon = True
+        bg.start()
+
+        exit_code = executor_shell.wait()
+        os._exit(exit_code)
+
+    # Close unused file descriptors to enforce PIPE behavior.
+    os.close(r)
+    os.close(stdout_w)
+    os.close(stderr_w)
+
+    # Redirect command stdout & stderr to provided streams or sys.stdout/sys.stderr.
+    # This is useful for Jupyter Notebook that uses custom sys.stdout/sys.stderr or
```

Comment:
Notebook --> Notebooks
uses --> use

Reply:
"Jupyter Notebook" is a name of software